nexus/src/db/datastore.rs

```diff
@@ -2924,6 +2926,65 @@ impl DataStore {
             .await
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
+
+    pub async fn update_available_artifact_upsert(
+        &self,
+        artifact: UpdateAvailableArtifact,
+    ) -> CreateResult<UpdateAvailableArtifact> {
+        use db::schema::update_available_artifact::dsl;
+        diesel::insert_into(dsl::update_available_artifact)
+            .values(artifact.clone())
+            .on_conflict((dsl::name, dsl::version, dsl::kind))
+            .do_update()
+            .set(artifact.clone())
+            .returning(UpdateAvailableArtifact::as_returning())
+            .get_result_async(self.pool())
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+    }
+
+    pub async fn update_available_artifact_hard_delete_outdated(
+        &self,
+        current_targets_role_version: i64,
+    ) -> DeleteResult {
+        // We use the `targets_role_version` column in the table to delete any old rows, keeping
+        // the table in sync with the current copy of artifacts.json.
+        use db::schema::update_available_artifact::dsl;
+        diesel::delete(dsl::update_available_artifact)
```

Comment:
artifacts.json should likely contain the list of all valid updates, including what Nexus can decide to downgrade itself to. (This might not scale good, and maybe the document format should be redesigned?)